nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -471,23 +481,122 @@ impl DataStore {
             .limit(1)
             .load_async(&*conn)
             .await?;
+
         if !old_resource.is_empty() {
             return Ok(old_resource[0].clone());
         }
 
-        let must_use_sleds: HashSet<&SledUuid> =
+        // Get a list of local storage disks attached to this instance
+        let local_storage_disks: Vec<LocalStorageDisk> = self
+            .instance_list_disks_impl_unauth(
+                &conn,
+                instance_id.into_untyped_uuid(),
+                &PaginatedBy::Name(DataPageParams {
+                    marker: None,
+                    direction: dropshot::PaginationOrder::Ascending,
+                    limit: std::num::NonZeroU32::new(MAX_DISKS_PER_INSTANCE)
+                        .unwrap(),
+                }),
+            )
+            .await?
+            .into_iter()
+            .filter_map(|disk| match disk {
+                db::datastore::Disk::LocalStorage(disk) => Some(disk),
+                db::datastore::Disk::Crucible(_) => None,
+            })
+            .collect();
+
+        // Gather constraints for the instance placement, starting with the
+        // arguments to this function.
+        let maybe_must_use_sleds: Option<HashSet<SledUuid>> =
             if let Some(must_select_from) = constraints.must_select_from() {
-                must_select_from.into_iter().collect()
+                Some(must_select_from.into_iter().cloned().collect())
             } else {
-                HashSet::default()
+                None
             };
 
+        // If any local storage disks have been allocated already, then this
+        // constraints VMM placement and where other unallocated local storage
+        // must be.
+        let maybe_must_use_sleds = if !local_storage_disks.is_empty() {
+            // Any local storage disk that was allocated already will have a
+            // sled_id.
+            let local_storage_allocation_sleds: HashSet<SledUuid> =
+                local_storage_disks
+                    .iter()
+                    .filter_map(|disk| {
+                        disk.local_storage_dataset_allocation
+                            .as_ref()
+                            .map(|allocation| allocation.sled_id())
+                    })
+                    .collect();
+
+            if local_storage_allocation_sleds.is_empty() {
+                // None of this instance's local storage disks have been
+                // allocated yet.
+                maybe_must_use_sleds
+            } else {
+                if local_storage_allocation_sleds.len() != 1 {
+                    // It's an error for multiple sleds to host local storage
+                    // disks, that makes no sense!
```

Comment:
Would this reflect a failure to uphold some constraint? Or is this a scenario we could encounter... somehow?

Reply:
Yes, but it's also a scenario that client requests can place us in if a user detaches a local storage disk with an allocation that's on sled A, then attaches it to an instance that has attached local storage with existing allocations on sled B. I've put a comment in c67d8e9668919136aae567d981fa936f354fb18e and changed this error to a conflict instead. There's some follow up work to do here.